Paywalls: Add initial Template2 UI and using colors and texts in paywall

tonidero opened the pull request:
### Description
Mostly a bunch of cleanup and improving the UI in this PR.
- Start using colors for texts and buttons.
- Start using device locale
- Move code to extensions for reusability
- Move paywall view UI to `Template2`. First template we're working on
- Improve Template2 to look closer to actual UI.
- Extract conversion from `Offering` to `PaywallViewState` for easier testing and reusability
- Make sure visibility of elements is correct.
![image](https://github.com/RevenueCat/purchases-android/assets/808417/b6b06544-5e53-4e46-b1e0-e14e6fe62694)

tonidero commented on ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/PaywallViewState.kt:
```diff
@@ -1,9 +1,17 @@
     object Loading : PaywallViewState()
     data class Error(val errorMessage: String) : PaywallViewState()
-    data class Loaded(val offering: Offering) : PaywallViewState()
+    data class Template2(
```
For now, we only have `Template2`. But ideally the UI is just driven from this state

NachoSoto replied:
There's only one template so it's fine for now, but this is going to be duplicated for every template.

I'd recommend creating a `PaywallTemplate` `enum` and changing this to just `Template` containing a `PaywallTemplate`. Then you can switch over that template enum to determine which view to display.

Does that make sense?

tonidero replied:
Yeah, that makes sense. We can do that refactor in future PRs 👍

tonidero commented on ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/templates/template2/Template2.kt:
```diff
@@ -0,0 +1,160 @@
+        val localizedConfig = state.paywallData.localizedConfig()
+        Text(
+            style = MaterialTheme.typography.displaySmall,
+            fontWeight = FontWeight.Black,
```
I had to change the font weight to make it closer to what we have in iOS/Dashboard. Will probably still require refinement but at least it's closer.

NachoSoto replied:
I'm surprised the title is "displaySmall"? 😅

tonidero replied:
Yeah, I was also surprised, but "displayed" styles are all huge 😅

tonidero commented on ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/templates/template2/Template2.kt:
```diff
@@ -0,0 +1,160 @@
+}
+
+@Composable
+private fun PurchaseButton(state: PaywallViewState.Template2, viewModel: PaywallViewModel) {
```
Some of these can probably get extracted once we start working on other Templates.

NachoSoto replied:
👍🏻 

Not necessary now but yeah definitely.

NachoSoto replied:
Also per the suggested refactor, that state would become `PaywallViewState.Template` instead which makes more sense.

tonidero commented on ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/templates/template2/Template2.kt:
```diff
@@ -0,0 +1,160 @@
+        colors.accent2Color to colors.accent1Color
+    } else {
+        // TODO-PAYWALLS: Find correct background unselected color
+        colors.backgroundColor to colors.text1Color
```
I didn't find what color to use for the unselected package button... Seems like in iOS we just use the default? Or maybe I missed it @NachoSoto

NachoSoto replied:
The logic is a bit complex, but TLDR it's no background colors:
```swift
if selected {
    self.roundedRectangle
        .foregroundColor(self.selectedBackgroundColor)
} else {
    if self.configuration.backgroundImageURLToDisplay != nil {
        // Blur background if there is a background image.
        self.roundedRectangle
            .foregroundStyle(.thinMaterial)
    } else {
        // Otherwise the text should have enough contrast with the selected background color.
        EmptyView()
    }
}
```

You can improve this in an upcoming PR, that's basically just polish :)

tonidero commented on ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/templates/template2/Template2.kt:
```diff
@@ -0,0 +1,160 @@
+    viewModel: PaywallViewModel,
+) {
+    val colors = state.paywallData.getColors()
+    val isSelected = aPackage == state.selectedPackage
```
This can probably be optimized with `remember` or `derivedStateOf`... But for now, I'm working on getting it to work and we can optimize later if needed.

NachoSoto replied:
Oh nice

tonidero commented on ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/templates/template2/Template2.kt:
```diff
@@ -0,0 +1,160 @@
+    ) {
+        // TODO-PAYWALLS: Replace with correct image
+        val drawable = LocalContext.current.packageManager.getApplicationIcon(LocalContext.current.packageName)
+        Image(drawable.toBitmap(config = Bitmap.Config.ARGB_8888).asImageBitmap(), contentDescription = "")
```
For now, I'm just using the App Icon. Once we implement a mechanism to actually use images from the dashboard we can use it.

tonidero commented on ui/revenuecatui/src/main/res/values/strings.xml:
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <string name="restore_purchases">Restore purchases</string>
```
I've only added the default (english) translation right now, but will need to add other translations to test localization properly for hardcoded strings.

NachoSoto replied:
👍🏻

tonidero commented on ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/PaywallViewModel.kt:
```diff
@@ -36,15 +45,49 @@ internal class PaywallViewModelImpl(offering: Offering?) : ViewModel(), PaywallV
+                Log.i(TAG, "Purchased package: ${purchaseResult.storeTransaction}")
             } catch (e: PurchasesException) {
-                Log.e("PaywallTester", "Error purchasing package: $e")
+                Log.e(TAG, "Error purchasing package: $e")
```
We should probably improve our logging mechanisms for the new library soon, to avoid this repetition and provide more details.

tonidero replied:
Done in #1233

NachoSoto commented on purchases/src/main/kotlin/com/revenuecat/purchases/paywalls/PaywallColor.kt:
```diff
@@ -24,6 +25,12 @@ data class PaywallColor(
+     * The color converted to a ColorInt
+     */
+    @ColorInt
+    val colorInt: Int = Color.parseColor(stringRepresentation)
```
I was thinking `intRepresentation` might make more sense, but if `@ColorInt` is an Android thing then maybe this name makes sense.

tonidero replied:
Well... I think either could work... But yeah, I kinda feel `colorInt` makes more sense, since it represents that... Lmk if you feel strongly though.

NachoSoto commented on ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/ContextExtensions.kt:
```diff
@@ -0,0 +1,16 @@
+import android.content.Context
+import android.content.ContextWrapper
+
+internal fun Context.getActivity(): Activity? {
```
Can you add a comment to explain what this does?

NachoSoto commented on ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/InternalPaywallView.kt:
```diff
@@ -1,83 +1,33 @@
+        is PaywallViewState.Error -> {
+            Text(text = "Error: ${state.errorMessage}")
+        }
+        is PaywallViewState.Template2 -> {
```
We'll probably want to extract this but I guess it's fine here for now.

NachoSoto commented on ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/PaywallDataExtensions.kt:
```diff
@@ -0,0 +1,51 @@
+@Composable
+@ReadOnlyComposable
+internal fun PaywallData.getColors(): PaywallData.Configuration.Colors {
+    return config.colors.dark?.takeIf { isSystemInDarkTheme() } ?: config.colors.light
```
`isSystemInDarkTheme` a static function? Is it possible to override that for snapshot tests for example?

tonidero replied:
Well, it's a composable function provided by Compose. I believe it can be overriden by hardcoding the light/night mode for the context, but for snapshot testing, we can configure the theme we're using so we can do snapshot tests for dark themes.

NachoSoto commented on ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/PaywallDataExtensions.kt:
```diff
@@ -0,0 +1,51 @@
+    val locale = getLocale()
+    val configForLocale = configForLocale(locale)
+    if (configForLocale == null) {
+        Log.e("RevenueCatUI", "No configuration found for locale $locale")
```
We'll probably want to extract this to avoid hardcoding `RevenueCatUI` everywhere, right?

tonidero replied:
Right, did that in a followup PR #1233

NachoSoto commented on ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/PaywallDataExtensions.kt:
```diff
@@ -0,0 +1,51 @@
+@ReadOnlyComposable
+private fun getLocale(): Locale {
+    val configuration = LocalConfiguration.current
+    return ConfigurationCompat.getLocales(configuration).get(0) ?: LocaleListCompat.getDefault()[0] ?: Locale.ENGLISH
```
Nice, so there is a list of configured locales, that's great! We can do the same we did an iOS and find the first one that's available:
https://github.com/RevenueCat/purchases-ios/blob/paywalls/Sources/Paywalls/PaywallData%2BLocalization.swift#L21-L26

NachoSoto replied:
Just added a Liner for this.

NachoSoto commented on purchases/src/main/kotlin/com/revenuecat/purchases/paywalls/PaywallColor.kt:
```diff
@@ -24,6 +25,12 @@ data class PaywallColor(
     val underlyingColor: Color?,
 ) {
+    /**
+     * The color converted to a ColorInt
```
Maybe
```suggestion
     * The color converted to a @ColorInt to be used by Compose
```

or something like that?

tonidero replied:
I believe the part of `to be used by Compose`, is not necessary since this is basically an int that can be used in different places, not only for Compose. We can add the `@` though, to make it clearer that this `int` represents a `@ColorInt`.

NachoSoto commented on ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/PaywallDataExtensions.kt:
```diff
@@ -0,0 +1,51 @@
+    get() = Color(callToActionForeground.colorInt)
+val PaywallData.Configuration.Colors.accent1Color: Color
+    get() = accent1?.let { Color(it.colorInt) } ?: callToActionForegroundColor
+val PaywallData.Configuration.Colors.accent2Color: Color
```
Nice

NachoSoto commented on ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/PaywallViewModel.kt:
```diff
@@ -19,14 +20,22 @@ import kotlinx.coroutines.launch
+    offering: Offering?,
+) : ViewModel(), PaywallViewModel {
+    companion object {
+        const val TAG = "RevenueCatUI"
```
Oh looks like this is defined here. Should this be extracted for other classes too?

tonidero replied:
Ah right yeah. We should have it in a single place. And also extract all the logging to a separate place 👍

tonidero replied:
Done in #1233

NachoSoto commented on ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/templates/template2/Template2.kt:
```diff
@@ -0,0 +1,160 @@
+package com.revenuecat.purchases.ui.revenuecatui.templates.template2
```
Nit, but I wonder if this needs its own folder or it can be inside `templates` instead?

tonidero replied:
Moved to the templates package in #1235

NachoSoto commented on ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/templates/template2/Template2.kt:
```diff
@@ -0,0 +1,160 @@
+                contentColor = colors.callToActionForegroundColor,
+            ),
+        ) {
+            Text(text = state.paywallData.localizedConfig()?.callToAction ?: "")
```
For the future: this will need to display `callToActionWithOffer` if applicable.

NachoSoto commented on ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/templates/template2/Template2.kt:
```diff
@@ -0,0 +1,160 @@
+    val border = if (isSelected) null else BorderStroke(2.dp, colors.text1Color)
+    Button(
+        modifier = Modifier.fillMaxWidth(),
+        onClick = { viewModel.selectPackage(aPackage) },
```
Love it

NachoSoto commented on ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/OfferingToStateMapper.kt:
```diff
@@ -0,0 +1,21 @@
+        paywallData = paywallData,
+        packages = packages,
+        selectedPackage = packages.first(), // TODO-PAYWALLS: Use correct default package
+        shouldShowRestorePurchasesButton = paywallData.config.displayRestorePurchases,
+        shouldShowTermsButton = paywallData.config.termsOfServiceURL != null,
+        shouldShowPrivacyPolicyButton = paywallData.config.privacyURL != null,
```
I wonder if these are needed here. The state already has `PaywallData` so the future `Footer` "compoment"/"view" whatever can derive it from its config?

tonidero replied:
Right, it could. I mostly wanted to abstract as much logic as possible away from the composables. Ideally, the state has everything it needs to render without making any decisions (we are not there yet)

NachoSoto reviewed (commented):
Awesome progress!!

tonidero commented:
I addressed some comments in the follow PRs. And I agree, we need to improve the architecture around how the data is passed to the composables for each template... But we can iterate on that on future PRs. Lmk if you have other questions @NachoSoto

NachoSoto commented on ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/ContextExtensions.kt:
```diff
@@ -0,0 +1,21 @@
+import android.content.ContextWrapper
+
+/**
+ * This function gets the activity from a given context. Most times, the context itself will be
```
Super nit, but I recommend not starting docs with stuff like "this function...". That's implied since it's document it 😅
```suggestion
 * Returns the activity from a given context. Most times, the context itself will be
```
